### PR TITLE
Check for the PKRU bit in XCR0

### LINKED ISSUE #2655: `PTRACE_SETREGSET` fails
https://github.com/mozilla/rr/issues/2034#issuecomment-691349725
> now I'm getting Assertion !errno' failed to hold. ptrace(PTRACE_SETREGSET, 110711, addr=0x202, data=0x7ffd971a9170) failed with errno 22`

https://github.com/mozilla/rr/issues/2034#issuecomment-691350469
> Happens every time for me (even, say just `rr record ls`).

It's not happening for me (5.7.15-100.fc31.x86_64). What kernel versions are you on @glandium @Keno  ?

### PR DESCRIPTION
**Pull request by glandium**
Fixes #2655

### CODE REVIEW

File: src/ExtraRegisters.cc [hunk 1]
```diff
@@ -596,10 +596,12 @@ void ExtraRegisters::reset() {
       * Avoid this issue by setting the bit if the feature is supported by the
       * CPU.
       */
-      const XSaveLayout& native_layout = xsave_native_layout();
       uint64_t pkru_bit = uint64_t(1) << xsave_feature_pkru;
-      if (native_layout.supported_feature_bits & pkru_bit) {
-        xinuse |= pkru_bit;
+      if (xcr0() & pkru_bit) {
+        const XSaveLayout& native_layout = xsave_native_layout();
+        if (native_layout.supported_feature_bits & pkru_bit) {
```
Comment (rocallahan) [inline src/ExtraRegisters.cc:602]: I don't think we need the native_layout check as well, do we? If the bit is enabled in XCR0 it had better be enabled in CPUID.

**Review by rocallahan — CHANGES_REQUESTED** [review]

File: src/ExtraRegisters.cc (hunk 1 above)
Comment (glandium) [reply src/ExtraRegisters.cc:602]: But the native_layout is not directly related to cpuid, is it?

Comment (rocallahan) [reply src/ExtraRegisters.cc:602]: `xsave_native_layout().supported_feature_bits` is basically just the result of `cpuid(CPUID_GETXSAVE, 0)`. We could probably just call the latter instead, if we wanted to, but I think we can just get rid of it.

Comment (rocallahan) [reply src/ExtraRegisters.cc:602]: The SDM says, for CPUID EAX=0x0d (CPUID_GETXSAVE), ECX=0:
> EAX Bits 31 - 00: Reports the supported bits of the lower 32 bits of XCR0. XCR0[n] can be set to 1 only if EAX[n] is 1.
So if we find that the PKRU bit is set in XCR0 but it is not set in CPUID_GETXSAVE/0 (i.e. `xsave_native_layout().supported_feature_bits`), then something has gone terribly wrong.

Comment (glandium) [reply src/ExtraRegisters.cc:602]: Fair enough.

**Pull request merged**